text/0005-application-formalization.md

```diff
@@ -0,0 +1,194 @@
+| uiModuleDescriptors | ModuleDescriptor[] | Module descriptors for the frontend modules/plugins comprising the application | Read-only. See [ModuleDescriptor Schema](https://github.com/folio-org/okapi/blob/master/okapi-core/src/main/raml/ModuleDescriptor.json) |
+| metadata          | Metadata     | System-generated record metadata                                         | Read-only. See [Metadata Schema] (https://github.com/folio-org/raml/blob/master/schemas/metadata.schema) |
+
+**Dependency Properties**
+| Property | Type    | Description                                                          | Notes                            |
+| ---------| ------- | -------------------------------------------------------------------- | -------------------------------- |
+| id       | String  | Identifier for the module conforming to pattern: `{name}-{version}`  | e.g. "mod-orders-storage-13.5.0" |
+| name     | String  | Name of the module                                                   | e.g. "mod-orders"                |
+| version  | String  | Version of the module                                                | e.g. "13.5.0"                    |
+
+**ModuleId Properties**
+| Property | Type    | Description                                                          | Notes                            |
+| -------- | ------- | -------------------------------------------------------------------- | -------------------------------- |
+| id       | String  | Identifier for the module conforming to pattern: `{name}-{version}`  | e.g. "mod-orders-storage-13.5.0" |
+| name     | String  | Name of the module                                                   | e.g. "mod-orders"                |
+| version  | String  | Version of the module                                                | e.g. "13.5.0"                    |
+| url      | String  | URL pointing to the module descriptor for this module                | e.g. "https://folio-registry.dev.folio.org/_/proxy/modules/mod-orders-storage-13.5.0" |
+
+Example:
+```
+{
+  "id": "app-gobi-0.0.1",
+  "name": "app-gobi",
+  "version": "0.0.1",
+  "description": "Application facilitating the placement of orders via GOBI and other compatible vendors",
+  "dependencies": [
+    {
+      "id": "app-orders-0.0.1",
+      "name": "app-orders",
+      "version": "^0.0.1"
+    },
+    {
+      "id": "app-organizations-0.0.1",
+      "name": "app-organizations",
+      "version": "^0.0.1"
+    }
+  ],
+  "modules": [
+    {
+      "id": "mod-gobi-2.6.0",
+      "name": "mod-gobi",
+      "version": "2.6.0",
+	  "url": "https://folio-registry.dev.folio.org/_/proxy/modules/mod-gobi-2.6.0"
+    }
+  ],
+  "uiModules": [
+    {
+      "id": "folio_gobi-settings-2.0.0",
+      "name": "folio_gobi-settings",
+      "version": "2.0.0",
+	  "url": "https://folio-registry.dev.folio.org/_/proxy/modules/folio_gobi-settings-2.0.0"
+    }
+  ]
+}
+```
```

Comment:
Like module descriptors, application descriptors should have a launch descriptor, which includes a list of all environment variables that the operator can pass to the module launcher (this might be Okapi, Kubernetes or something else). Environment variables should be listed with their default values.

An application manager will pick up this list, override default values with a set of values supplied by the operator (they might be Okapi's environment variables or a config map, a properties file or secrets used by a Kubernetes deployment, or something else) and pass the chosen values to the module launcher of each module within the application.

Environment variables can thus be defined in one central place for each application.
This will make FOLIO deployment considerably easier, because much less times the operator needs to check the validity of an environment variables list as it is now with 60+ different modules.

Reply:
@inkuss 

> Like module descriptors, application descriptors should have a launch descriptor, which includes a list of all environment variables that the operator can pass to the module launcher (this might be Okapi, Kubernetes or something else). Environment variables should be listed with their default values.

As I understand it, applications are not a deployable thing, thus they do not have environment variables or other configuration

Reply:
@inkuss This suggests to me that you expect applications to be a deployable unit or for the applications manager to deploy the application or it's constituent modules

That is not my understanding of how applications or how the application management will work

Reply:
Right. I expect the application manager to be able to deploy applications. If it won't be able to, the whole formalism will not be a great help for system administrators. It will just reduce the number of dependencies (but those are being resolved anyway when I register the modules to Okapi).

Reply:
> I expect the application manager to be able to deploy applications

I could easily be wrong

My understanding from the beginning (much like with the other app store initiatives) is that module deployment is out of scope because it varies significantly by environment and the application manager has no generalised / common way of accessing the deployment tooling that is specific to that environment

Reply:
I just checked, I think the [application management](https://github.com/folio-org/rfcs/blob/0006-application-management/text/0006-application-management.md) RFC is where this topic will be covered and this includes module deployment in the scope

Reply:
Just because we have application descriptors does not mean that module descriptors are going away.  They continue to exist, complete with launch descriptors.  I personally do see any value in duplicating the launch descriptor information in the application descriptor.

Reply:
I agree with @marcjohnson-kint that the discussion about deployment is probably better saved for the application mgmt RFC.

Reply:
I agree that the question of deployment by an application manager should be part of 0006-application-management.md. But there is no pull request for the Draft Refinement stage for this rfc, yet. - I still think that not duplicating the launch descriptor information into the application decriptor is already a design decision which needs to be made here (at this pr).

Reply:
Can I open an rfc for  0006-application-management.md draft refinement case of who will do so ? If I have put my concerns there I think this conversation can be resolved.

Reply:
0006-application-management.md is still being written.  My suggestion is to simply raise this concern again once that RFC is in the DRAFT REFINEMENT stage.

> I still think that not duplicating the launch descriptor information into the application decriptor is already a design decision which needs to be made here (at this pr).

I've made an adjustment in the scope section indicating that application deployment is out of scope.

I've also added a note below the table of Application Descriptor properties clarifying that module descriptors are not being replaced or retired, so it's not required to duplicate or move info like the launch descriptor section into the application descriptors.

Does this suffice @inkuss ?

Reply:
Thank you, I will raise it again in the DRAFT REFINEMENT stage of 0006-application-management.md.